agent/structs/config_entry.go

```diff
@@ -0,0 +1,226 @@
+	EnableTagOverride bool
+
+	// Non script/docker checks only
+	Check  *HealthCheck
+	Checks HealthChecks
+
+	// Kind is allowed to accommodate non-sidecar proxies but it will be an error
+	// if they also set Connect.DestinationServiceID since sidecars are
+	// configured via their associated service's config.
+	Kind ServiceKind
+
+	// Only DestinationServiceName and Config are supported.
+	Proxy ConnectProxyConfig
+
+	Connect ServiceConnect
+
+	Weights Weights
+}
+
+// ProxyConfigEntry is the top-level struct for global proxy configuration defaults.
+type ProxyConfigEntry struct {
+	Kind   string
+	Name   string
+	Config map[string]interface{}
+
+	RaftIndex
+}
+
+func (e *ProxyConfigEntry) GetKind() string {
+	return ProxyDefaults
+}
+
+func (e *ProxyConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+
+	return e.Name
+}
+
+func (e *ProxyConfigEntry) Normalize() error {
+	if e == nil {
+		return fmt.Errorf("config entry is nil")
+	}
+
+	e.Kind = ProxyDefaults
+
+	return nil
+}
+
+func (e *ProxyConfigEntry) Validate() error {
+	if e == nil {
+		return fmt.Errorf("config entry is nil")
+	}
+
+	if e.Name != ProxyConfigGlobal {
+		return fmt.Errorf("invalid name (%q), only %q is supported", e.Name, ProxyConfigGlobal)
+	}
+
+	return nil
+}
+
+func (e *ProxyConfigEntry) GetRaftIndex() *RaftIndex {
+	if e == nil {
+		return &RaftIndex{}
+	}
+
+	return &e.RaftIndex
+}
+
+type ConfigEntryOp string
+
+const (
+	ConfigEntryUpsert ConfigEntryOp = "upsert"
+	ConfigEntryDelete ConfigEntryOp = "delete"
+)
+
+type ConfigEntryRequest struct {
+	Op    ConfigEntryOp
+	Entry ConfigEntry
+}
+
+func (r *ConfigEntryRequest) MarshalBinary() (data []byte, err error) {
+	// bs will grow if needed but allocate enough to avoid reallocation in common
+	// case.
+	bs := make([]byte, 128)
+	enc := codec.NewEncoderBytes(&bs, msgpackHandle)
+	// Encode kind first
+	err = enc.Encode(r.Entry.GetKind())
+	if err != nil {
+		return nil, err
+	}
+	// Then actual value using alias trick to avoid infinite recursion
+	type Alias ConfigEntryRequest
+	err = enc.Encode(struct {
+		*Alias
+	}{
+		Alias: (*Alias)(r),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return bs, nil
+}
+
+func (r *ConfigEntryRequest) UnmarshalBinary(data []byte) error {
+	// First decode the kind prefix
+	var kind string
+	dec := codec.NewDecoderBytes(data, msgpackHandle)
+	if err := dec.Decode(&kind); err != nil {
+		return err
+	}
+
+	// Then decode the real thing with appropriate kind of ConfigEntry
+	r.Entry = makeConfigEntry(kind)
+
+	// Alias juggling to prevent infinite recursive calls back to this decode
+	// method.
+	type Alias ConfigEntryRequest
+	as := struct {
+		*Alias
+	}{
+		Alias: (*Alias)(r),
+	}
+	if err := dec.Decode(&as); err != nil {
+		return err
+	}
+	return nil
+}
+
+func makeConfigEntry(kind string) ConfigEntry {
+	switch kind {
+	case ServiceDefaults:
+		return &ServiceConfigEntry{}
+	case ProxyDefaults:
+		return &ProxyConfigEntry{}
+	default:
+		panic("invalid kind")
```

Comment:
We should maybe return error here instead of panic. In theory this could be called during normal raft operations - log replication etc and we should probably leave it up to the call site to decide if the unknown type is worth panicing for. For Raft replication maybe it is in the same was as unknown log entries but I don't think we should make that decision final here - the `UnmarshalBinary` has a way to return an error normally so I suggest we just do that here to allow the call site to decide if it should gracefully continue.